SpaceService: Messaging between WorkplaceService

mekk1t opened the pull request.

mekk1t committed:
Update OfficeServiceClient

1. Moved Client Interface to the respective folder;
2. Added Console Text for acknowledgment of connecting to the queue.

mekk1t committed:
Implemented WorkplaceServiceClient

Implemented messaging between SpaceService & WorkplaceService.

-- Created a class and an interface called " [I]WorkplaceServiceClient";
-- Created a "WorkplaceResponse" model;
-- Added "WorkplaceResponse Workplaces" property in the "SpaceResponse" model;
-- Registered WorkplaceServiceClient in Startup.cs;
-- Implemented messaging in the SpaceController.

mekk1t committed:
RabbitMQ: Cloud

Set up connection to the cloud server via AMQP-protocol.

mekk1t committed:
Database: DbContext & Models Redefined

1. Deleted DbOffice
2. Deleted links on the other end of the relationships
3. DbSpace: renamed properties
4. DbContext: removed OnConfiguring, deleted modelbuilder for DbOffce, redefined relationships via Fluent API

mekk1t committed:
Update SpaceResponse.cs

Corrected an output model: added the "SpaceName" property and removed the  "OfficeGuid" property.

aatroshcenko commented on OfficeMapApi/SpaceService/SpaceService/Clients/WorkplaceServiceClient.cs:
```diff
@@ -0,0 +1,84 @@
+            return item;
+        }
+
+        public void Close()
+        {
+            rabbitMQPersistentConnection.Close();
+        }
```
а гду ты вот этот метод используешь?

aatroshcenko commented on OfficeMapApi/SpaceService/SpaceService/Clients/WorkplaceServiceClient.cs:
```diff
@@ -0,0 +1,84 @@
+            rabbitMQPersistentConnection.Close();
+        }
+
+        public void Dispose()
+        {
+            rabbitMQPersistentConnection.Dispose();
+        }
```
надо добвить интерфес IDispose

aatroshcenko commented on OfficeMapApi/SpaceService/SpaceService/Controllers/SpaceController.cs:
```diff
@@ -37,13 +40,16 @@ public async Task<ActionResult> GetSpaces([FromRoute] Guid officeGuid)
                 {
+                    var _response = workplaceServiceClient.GetWorkplacesAsync(space.SpaceGuid).Result;
+                    space.Workplaces = _response;
                     space.OfficeGuid = filter.OfficeGuid;
                     InnerSpaces(space.Spaces, officeGuid);
```
а почемы мы назначаем officeGuid, мы же должны его получить из базы

mekk1t replied:
Э-э-э, я видимо это уже успел поправить; у себя я не нахожу таких строк.

Тем не менее, как мы получим гуид из базы, если он у нас подается в адресной строке? Допустим, есть факт существования офиса -- так мы в случае response != null назначим/возьмем прямо из входного аргумента.

aatroshcenko commented on OfficeMapApi/SpaceService/SpaceService/Controllers/SpaceController.cs:
```diff
@@ -79,7 +85,10 @@ public async Task<ActionResult> GetSpaces([FromRoute] Guid officeGuid)
             {
+
                 var result = await spaceService.GetAsync(spaceGuid);
+                var _response = workplaceServiceClient.GetWorkplacesAsync(spaceGuid).Result;
```
надо асинхронно вызывать через await

aatroshcenko commented on OfficeMapApi/SpaceService/SpaceService/Controllers/SpaceController.cs:
```diff
@@ -37,13 +40,16 @@ public async Task<ActionResult> GetSpaces([FromRoute] Guid officeGuid)
-                foreach (var space in spaces)
+                var result = await spaceService.FindAsync(filter);
+                
+                foreach (var space in result)
                 {
+                    var _response = workplaceServiceClient.GetWorkplacesAsync(space.SpaceGuid).Result;
+                    space.Workplaces = _response;
                     space.OfficeGuid = filter.OfficeGuid;
```
можно обернуть в Parallel.ForEach - эту штука позволит вызвать все асинхронно

mekk1t committed:
RabbitMQ: Client to Producer

Changed OfficeService/WorkplaceServiceClient.cs to one Producer.cs
-- it creates an exchange "IDs" to the WS and OS queues. (WorkplaceService & OfficeService respectively)
-- the queues are binded to the exchange via BindingKeys "officeguid" and "spaceid"
-- callback queues are Guid-generated temporary ones. Once the message is delivered back to the Client/Producer, the queue is auto-deleted.

mekk1t committed:
Update Producer.cs

mekk1t committed:
Server & Client

-- Implemented messaging between SS, WS and OS
-- Created SpaceServiceServer to receive messages from WorkplaceService. Implementation of the exchange, queues and BindingKeys remains to be done.

mekk1t committed:
Splitted ServicesClient into 2 different Clients

Created 2 separate classes:
1. OfficeServiceClient;
2. WorkplaceServiceClient;

mekk1t committed:
Implemented Filter for IGet interface

Now we can filter our GET /{spaceGuid} data output.

mekk1t committed:
Deleted comments

mekk1t committed:
RabbitMQ: SpaceServiceServer

1. Created "SpaceServiceServer" class. Registered it in Startup.cs;
2. Changed WorkplaceResponse.WorkplaceGuid to "Guid" to match property from the WorkplaceService;
3. Created a "WorkplaceRequest" model;
4. Implemented workplaceServiceClient in SpaceController + deleted useless "InnerSpaces" method;

mekk1t committed:
Update SpaceService.csproj

Deleted "EFCore+" package

mekk1t committed:
Update IWorkplaceServiceClient.cs

Changed identity from Guid to Id

mekk1t committed:
Update Automapper

Deleted useless mappings

mekk1t committed:
Update SpaceService Server + Controller

-- Changed data from controller to data from officeServiceClient + spaceService
-- Added await to async methods in SpaceController [GET /{spaceGuid}]

mekk1t committed:
Dummy WorkplaceResponse

mekk1t committed:
Update Startup.cs

Removed Controller registration via services

aatroshcenko commented on OfficeMapApi/SpaceService/SpaceService/Controllers/SpaceController.cs:
```diff
@@ -68,22 +70,24 @@ public async Task<ActionResult> GetSpaces([FromRoute] Guid officeGuid)
-                return Ok(result);
-            }
+            SpaceFilter filter = new SpaceFilter(response.OfficeId, response.Guid);  
+            var result = await spaceService.GetAsync(spaceGuid, filter);
```
а почему мы передаем spaceguid и filter в GetAsync?

mekk1t replied:
Потому что надо добавить фильтр по OfficeId.

Я проверил так, что в запрос кинул гуид Офиса3, а гуид спейса -- из офиса1.

Ну, он мне и выдал помещение вообще из другого офиса :)

aatroshcenko commented on OfficeMapApi/SpaceService/SpaceService/Models/SpaceResponse.cs:
```diff
@@ -6,15 +6,13 @@ namespace SpaceService.Models
-
         public Guid SpaceGuid { get; set; }
-        public Guid OfficeGuid { get; set; }
+        public int SpaceId { get; set; }
```
почему тут появился SpaceId? фронт не должен знать о них

mekk1t replied:
Хм, не помню. Надо пересмотреть

aatroshcenko commented on OfficeMapApi/SpaceService/SpaceService/Repository/Interfaces/ISpaceRepository.cs:
```diff
@@ -11,7 +11,7 @@ public interface ISpaceRepository :
         IDelete<DbSpace>,
         IFind<DbSpace, SpaceFilter>,
-        IGet<DbSpace, Guid>,
+        IGet<DbSpace, SpaceFilter, Guid>,
```
зачем? почему в фильтр гуинд этот не добавить?

aatroshcenko commented on OfficeMapApi/SpaceService/SpaceService/Startup.cs:
```diff
@@ -42,13 +44,18 @@ public void ConfigureServices(IServiceCollection services)
             {
                 return new ConnectionFactory()
                 {
+                    Uri = new System.Uri(Configuration["CLOUDAMQP_URL"]),
                     HostName = Configuration["RabbitMQConnection"],
                     UserName = Configuration["RabbitMQUsername"],
                     Password = Configuration["RabbitMQPassword"]
                 };
```
используй конфигурации из Azure. посмотри как в develop это выглядит( в Program.cs, Startup.cs, appsetting.json)

mekk1t replied:
Да, внесу правки. Они уже есть в других сервисах, добавлю и тут.

mekk1t committed:
Azure Settings

-- Updated Startup, Program and appsettings.json to use Azure's config;
-- Installed 2 NuGet packages: jwtbearer and azureappconfiguration.

mekk1t committed:
RabbitMQ Space Model

Created a new model for messaging

mekk1t committed:
New Message Model

Created a message model called "SpaceRequest" which contains
int SpaceId.
Implemented it in the SpaceServiceServer that receives a space model from WorkplaceService.

mekk1t committed:
RabbitMQ: Logging

Added logging to the client and Connection classes.

mekk1t committed:
Update Files

-- renamed _response to space in SpaceController so that we understand what model we operate with;
-- removed unused data configuration from appsettings.json
